client/rpc-spec-v2/src/chain_head/chain_head.rs

```diff
@@ -0,0 +1,477 @@
+	};
+
+	let parent_rt = match client.runtime_version_at(parent) {
+		Ok(rt) => rt,
+		Err(err) => return to_event_err(err),
+	};
+
+	// Report the runtime version change.
+	if block_rt != parent_rt {
+		Some(RuntimeEvent::Valid(RuntimeVersionEvent { spec: block_rt }))
+	} else {
+		None
+	}
+}
+
+#[async_trait]
+impl<BE, Block, Client> ChainHeadApiServer<Block::Hash> for ChainHead<BE, Block, Client>
+where
+	Block: BlockT + 'static,
+	Block::Header: Unpin,
+	BE: Backend<Block> + 'static,
+	Client: BlockBackend<Block>
+		+ ExecutorProvider<Block>
+		+ HeaderBackend<Block>
+		+ BlockchainEvents<Block>
+		+ CallApiAt<Block>
+		+ StorageProvider<Block, BE>
+		+ 'static,
+{
+	fn chain_head_unstable_follow(
+		&self,
+		mut sink: SubscriptionSink,
+		runtime_updates: bool,
+	) -> SubscriptionResult {
+		let sub_id = self.accept_subscription(&mut sink)?;
+		// Keep track of the subscription.
+		self.subscriptions.insert_subscription(sub_id.clone());
+
+		let client = self.client.clone();
+		let subscriptions = self.subscriptions.clone();
+
+		let sub_id_import = sub_id.clone();
+		let stream_import = self
+			.client
+			.import_notification_stream()
+			.map(move |notification| {
+				let new_runtime = generate_runtime_event(
+					&client,
+					runtime_updates,
+					&BlockId::Hash(notification.hash),
+					Some(&BlockId::Hash(*notification.header.parent_hash())),
+				);
+
+				let _ = subscriptions.pin_block(&sub_id_import, notification.hash);
+
+				// Note: `Block::Hash` will serialize to hexadecimal encoded string.
+				let new_block = FollowEvent::NewBlock(NewBlock {
+					block_hash: notification.hash,
+					parent_block_hash: *notification.header.parent_hash(),
+					new_runtime,
+					runtime_updates,
+				});
+
+				if !notification.is_new_best {
+					return stream::iter(vec![new_block])
+				}
+
+				// If this is the new best block, then we need to generate two events.
+				let best_block = FollowEvent::BestBlockChanged(BestBlockChanged {
+					best_block_hash: notification.hash,
+				});
+				stream::iter(vec![new_block, best_block])
+			})
+			.flatten();
+
+		let subscriptions = self.subscriptions.clone();
+		let sub_id_finalized = sub_id.clone();
+
+		let stream_finalized =
+			self.client.finality_notification_stream().map(move |notification| {
+				// We might not receive all new blocks reports, also pin the block here.
+				let _ = subscriptions.pin_block(&sub_id_finalized, notification.hash);
+
+				// The tree route contains the exclusive path from the latest finalized block
+				// to the block reported by the notification. Ensure the finalized block is
+				// properly reported to that path.
+				let mut finalized_block_hashes =
+					notification.tree_route.iter().cloned().collect::<Vec<_>>();
+				finalized_block_hashes.push(notification.hash);
+
+				FollowEvent::Finalized(Finalized {
+					finalized_block_hashes,
+					pruned_block_hashes: notification.stale_heads.iter().cloned().collect(),
+				})
```

Comment:
I don't exactly know how the client works, but it might be possible that the `pruned_block_hashes` contains the current best block, in which case the best block should have changed to a different one. If that happens, you need to send a `FollowEvent::BestBlockChanged` before the `Finalized`.

Reply:
To rephrase:

Conceptually, the best block must never be pruned. If an update in the finality would lead you to prune the best block (which is totally possible), then you should change the best block to be a different block.

The client most likely does all of this at the same time: update the finalized block, prune the blocks that aren't a descend of it, and update the best block.
In the JSON-RPC API, if that happens, you have to send a `FollowEvent::BestBlockChanged` before the `FollowEvent::Finalized`, because I don't think it makes a difference whether you change the best block right before or at the same time, and it makes it I think easier to implement for the JSON-RPC client.

Reply:
I've handled this case by checking if the last reported block by the RPC is now reported as pruned by the finalization event. In that case, we also report a new `BestBlockChanged` obtained from the client's internal data `client.info().best_hash`. Let me know what you think of it, thanks!

Reply:
The approach is good, but you should also check whether the last reported block is an ancestor of the now finalized block.
A best block equal to the current finalized block is ok, but a best block that is an ancestor of the current finalized block is not ok.